[WPEPlatform] Introduce support for Android's AHardwareBuffer

aperezdc opened the pull request:
#### fd72faeef06a1b166d79872ae8d5c96182996c69
<pre>
[WPEPlatform] Introduce support for Android&apos;s AHardwareBuffer
<a href="https://bugs.webkit.org/show_bug.cgi?id=294317">https://bugs.webkit.org/show_bug.cgi?id=294317</a>

Reviewed by Carlos Garcia Campos.

Introduce a new WPEBufferAndroid subclass of WPEBuffer that wraps an
AHardwareBuffer instance. Similarly to WPEBufferDMABuf, it supports
setting rendering and release fences (as file descriptors). The
WPEBuffer.import_to_pixels virtual method is left unimplemented for
now, but it should be possible to provide the operation for buffers
created with AHARDWAREBUFFER_USAGE_CPU_READ_{RARELY,OFTEN} in the
future.

* Source/WebKit/Shared/glib/RendererBufferFormat.h:
* Source/WebKit/WPEPlatform/CMakeLists.txt:
* Source/WebKit/WPEPlatform/wpe/WPEBufferAndroid.cpp: Added.
(createImageEGL15):
(createImageKHRImageBase):
(wpeBufferAndroidDisposeEGLImageIfNeeded):
(wpeBufferAndroidDispose):
(wpeBufferAndroidImportToEGLImage):
(wpe_buffer_android_class_init):
(wpe_buffer_android_new):
(wpe_buffer_android_get_hardware_buffer):
(wpe_buffer_android_get_format):
* Source/WebKit/WPEPlatform/wpe/WPEBufferAndroid.h: Copied from Source/WebKit/WPEPlatform/wpe/WPEConfig.h.in.
* Source/WebKit/WPEPlatform/wpe/WPEConfig.h.in:
* Source/WebKit/WPEPlatform/wpe/wpe-platform.h:
* Source/cmake/OptionsWPE.cmake:

Canonical link: <a href="https://commits.webkit.org/298619@main">https://commits.webkit.org/298619@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/8351b30d51bfa2a487ffe7d6fc28cd62546b0516

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/116075 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/35736 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/26278 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/122131 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/66625 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/36430 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/44324 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/88184 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/66625 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/119023 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/29059 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/104157 "Passed tests") | [❌ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/68595 "Found 1 new API test failure: /WPE/TestSSL:/webkit/WebKitWebView/ephemeral-tls-errors (failure)") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/28174 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/65813 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/108184 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/98447 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/22402 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/125281 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/114603 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/42969 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/32257 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/96917 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/43334 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/100347 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/96701 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/24606 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/41973 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/19861 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/38915 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/42856 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/48448 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/143300 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/42323 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/36943 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/45658 "Built successfully") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/124/builds/44027 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/WPEBufferAndroid.cpp:
```diff
@@ -0,0 +1,349 @@
+WEBKIT_DEFINE_FINAL_TYPE(WPEBufferAndroid, wpe_buffer_android, WPE_TYPE_BUFFER, WPEBuffer)
+
+static EGLImage
+createImageEGL15(EGLDisplay display, AHardwareBuffer * ahb)
```
```suggestion
static EGLImage createImageEGL15(EGLDisplay display, AHardwareBuffer * ahb)
```

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/WPEBufferAndroid.cpp:
```diff
@@ -0,0 +1,349 @@
+}
+
+static EGLImage
+createImageKHRImageBase(EGLDisplay display, AHardwareBuffer* ahb)
```
```suggestion
static EGLImage createImageKHRImageBase(EGLDisplay display, AHardwareBuffer* ahb)
```

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/WPEBufferAndroid.cpp:
```diff
@@ -0,0 +1,349 @@
+createImageEGL15(EGLDisplay display, AHardwareBuffer * ahb)
+{
+    static constexpr std::array<EGLAttrib, 3> attributes = { EGL_IMAGE_PRESERVED, EGL_TRUE, EGL_NONE };
+    return eglCreateImage(display, EGL_NO_CONTEXT, EGL_NATIVE_BUFFER_ANDROID, ahb, attributes.data());
```
Is there always a current context in android? Because it's required by epoxy to find the symbols, but EGL image is display api that doesn't need a context.

aperezdc replied:
Good point. It looks like by chance we got a current context here, but it would be better to not rely on always having one and resolving the symbols with `epoxy_eglGetProcAddress()`&mdash;like it's done for the DMA-BUF buffer type. I'll change it.

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/WPEBufferAndroid.cpp:
```diff
@@ -0,0 +1,349 @@
+ *
+ * Create a new #WPEBufferAndroid for the given buffer.
+ *
+ * The reference cound of the @ahb will be incremented using
```
```suggestion
 * The reference count of the @ahb will be incremented using
```

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/WPEBufferAndroid.cpp:
```diff
@@ -0,0 +1,349 @@
+ *
+ * Returns: (transfer full): a #WPEBufferAndroid
+ */
+WPEBufferAndroid* wpe_buffer_android_new_take(WPEDisplay* display, AHardwareBuffer* ahb)
```
According to the other patch, this is the only constructor used, so I would remove the other one. I think it's better to not take the ownership, but simply take a reference and release it when done.

aperezdc replied:
Ok, let's keep only one, and call it just `wpe_buffer_android_new()` 👍🏼

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/WPEBufferAndroid.cpp:
```diff
@@ -0,0 +1,349 @@
+ * The fence will be used to wait before rendering the buffer. The
+ * buffer takes ownership of the file descriptor.
+ */
+void wpe_buffer_android_set_rendering_fence(WPEBufferAndroid* buffer, int fd)
```
The other patch sets the rendering fence, but the getter is never called, so I wonder if fences are actually handled. In DMABuf buffers this fences are only used when explicit sync is enabled, which means the fences are handled by wayland compositor or DRM. Is there API in android to deal with the rendering/release fences?

aperezdc replied:
Yes, there are APIs in Android that can use the fences. For example:

- A fence can be passed to [AHardwareBuffer_lock()](https://developer.android.com/ndk/reference/group/a-hardware-buffer#ahardwarebuffer_lock) to make it wait until the fence is signalled, then map the buffer (typically used to ensure rendering is done before reading pixel data).
- A fence can be passed to [ASurfaceTransaction_setBuffer()](https://developer.android.com/ndk/reference/group/native-activity#asurfacetransaction_setbuffer) so SurfaceFlinger (the compositor) can wait on it before using the buffer.
- We can install a completion callback with [ASurfaceTransaction_setOnComplete()](https://developer.android.com/ndk/reference/group/native-activity#asurfacetransaction_setoncomplete), and at the point the callback is invoked it is possible to reuse buffers and therefore that's the point at which we can signal the release fence to let WebKit reuse the buffer.

These are to be used in the code that lives in [wpe-android](https://github.com/Igalia/wpe-android/), which currently uses a legacy WPEBackend, and eventually will be migrated to WPEPlatform. The code in there will want to use the render and release fences associated to the `WPEBufferAndroid` instance.

carlosgcampos replied:
Then, maybe we should move the fences apis to the base class.

carlosgcampos replied:
https://github.com/WebKit/WebKit/pull/49194

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/WPEBufferAndroid.cpp:
```diff
@@ -0,0 +1,349 @@
+ * Returns: (transfer none): a valid #AHardwareBuffer pointer.
+ */
+AHardwareBuffer*
+wpe_buffer_android_get_hardware_buffer(WPEBufferAndroid* buffer)
```
```suggestion
AHardwareBuffer* wpe_buffer_android_get_hardware_buffer(WPEBufferAndroid* buffer)
```

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/WPEBufferAndroid.h:
```diff
@@ -0,0 +1,69 @@
+
+WPE_API AHardwareBuffer  *wpe_buffer_android_get_hardware_buffer  (WPEBufferAndroid *buffer);
+
+WPE_API guint32           wpe_buffer_android_get_format_fourcc    (WPEBufferAndroid *buffer);
```
I would use just format for consistency with DMABuf buffers, or maybe we want to change the DMABuf buffers to get_format_fourcc or just get_fourcc

aperezdc replied:
I'll go with `wpe_buffer_android_get_format()` 👍🏼

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/WPEBufferAndroid.cpp:
```diff
@@ -0,0 +1,349 @@
+ * Returns: a DRM FourCC code.
+ */
+guint32
+wpe_buffer_android_get_format_fourcc(WPEBufferAndroid* buffer)
```
```suggestion
guint32 wpe_buffer_android_get_format_fourcc(WPEBufferAndroid* buffer)
```

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/WPEBufferAndroid.cpp:
```diff
@@ -0,0 +1,231 @@
+    GObjectClass* objectClass = G_OBJECT_CLASS(bufferAndroidClass);
+    objectClass->dispose = wpeBufferAndroidDispose;
+
+    // TODO: Implement import_to_pixels vfunc.
```
```suggestion
    // FIXME: Implement import_to_pixels vfunc.
```

aperezdc replied:
Even better, I have created [an issue](https://bugs.webkit.org/show_bug.cgi?id=297316) for this, so it land with `FIXME(297316)` 😄

aperezdc committed:
[WPEPlatform] Introduce support for Android's AHardwareBuffer
https://bugs.webkit.org/show_bug.cgi?id=294317

Reviewed by Carlos Garcia Campos.

Introduce a new WPEBufferAndroid subclass of WPEBuffer that wraps an
AHardwareBuffer instance. Similarly to WPEBufferDMABuf, it supports
setting rendering and release fences (as file descriptors). The
WPEBuffer.import_to_pixels virtual method is left unimplemented for
now, but it should be possible to provide the operation for buffers
created with AHARDWAREBUFFER_USAGE_CPU_READ_{RARELY,OFTEN} in the
future.

* Source/WebKit/Shared/glib/RendererBufferFormat.h:
* Source/WebKit/WPEPlatform/CMakeLists.txt:
* Source/WebKit/WPEPlatform/wpe/WPEBufferAndroid.cpp: Added.
(createImageEGL15):
(createImageKHRImageBase):
(wpeBufferAndroidDisposeEGLImageIfNeeded):
(wpeBufferAndroidDispose):
(wpeBufferAndroidImportToEGLImage):
(wpe_buffer_android_class_init):
(wpe_buffer_android_new):
(wpe_buffer_android_get_hardware_buffer):
(wpe_buffer_android_get_format):
* Source/WebKit/WPEPlatform/wpe/WPEBufferAndroid.h: Copied from Source/WebKit/WPEPlatform/wpe/WPEConfig.h.in.
* Source/WebKit/WPEPlatform/wpe/WPEConfig.h.in:
* Source/WebKit/WPEPlatform/wpe/wpe-platform.h:
* Source/cmake/OptionsWPE.cmake:

Canonical link: https://commits.webkit.org/298619@main
